### PR TITLE
Forms in add resource

### PR DESCRIPTION
**Pull request by Lama64**

### CODE REVIEW

File: web/src/lib/components/ui/add-resource/add-resource.svelte [hunk 1]
```diff
@@ -1,125 +1,159 @@
...
-        {/if}
+    {#if names.includes('capacity')}
+      <div class="text-lg font-bold flex">Kapazität:</div>
+      <div class="flex flex-col gap-1 w-80">
+        <Form.Field {form} name="capacity">
+          <Form.Control let:attrs>
+            <Input
+              type="number"
+              min="0"
+              {...attrs}
```
Comment (danieldietzler) [inline web/src/lib/components/ui/add-resource/add-resource.svelte:84]: This should be the first prop on the input, as you want everything else here to be overrides

File: web/src/lib/components/ui/tags-multiple-select/tags-multiple-select.svelte [hunk 2]
```diff
@@ -2,17 +2,14 @@
...
   let selectedTags: Selected<string>[] = [];
-  export let selectedTagIds: unknown[];
-  export let entityTags: Tag[];
-
-  onMount(() => {
-    selectedTags = entityTags.map((tag) => ({ label: tag.name, value: tag.id }));
-    selectedTagIds = selectedTags.map((tag) => tag.value);
-  });
+  export let selectedTagIds: string[];
+  selectedTags = tags
```
Comment (danieldietzler) [inline web/src/lib/components/ui/tags-multiple-select/tags-multiple-select.svelte:9]: Two minor things: (1) exports should be in one batch at the top. (2) you can just immediately assign `tags.filter().map()` to the `selectedTags` field. Also, I think this should be reactive, so `$: selectedTags = tags.filter().map()`, right?

File: web/src/routes/admin/(resources)/grades/[id]/+page.svelte [hunk 3]
```diff
@@ -2,45 +2,26 @@
...
 
-  async function create(values: string[], tagIds: string[]) {
+  async function update(formData: Record<string, string | number | string[]>) {
     let gradeRequestDto: GradeRequestDto = {
-      name: values[0],
-      studentGroupIds: [],
-      tagIds,
+      name: String(formData.name),
+      studentGroupIds: formData.studentGroups as string[],
+      tagIds: formData.tags as string[],
```
Comment (danieldietzler) [inline web/src/routes/admin/(resources)/grades/[id]/+page.svelte:14]: We could probably type `formData` stronger such that we don't need those casts. Not sure if it's worth it though

File: web/src/routes/admin/(resources)/grades/[id]/+page.svelte [hunk 4]
```diff
@@ -2,45 +2,26 @@
...
-      name: values[0],
-      studentGroupIds: [],
-      tagIds,
-    };
-    try {
-      await updateGrade(grade.id, gradeRequestDto);
+      const _ =
+        formData.id === 'new'
+          ? await createGrade(gradeRequestDto)
+          : await updateGrade(String(formData.id), gradeRequestDto);
```
Comment (danieldietzler) [inline web/src/routes/admin/(resources)/grades/[id]/+page.svelte:20]: Don't. Just use an if lol

File: web/src/routes/admin/(resources)/rooms/[id]/+page.svelte [hunk 5]
```diff
@@ -1,48 +1,28 @@
...
-      buildingName: values[1],
-      capacity: Number(values[2]),
-      tagIds,
-    };
-    try {
-      await updateRoom(room.id, roomRequestDto);
+      const _ =
+        formData.id === 'new'
+          ? await createRoom(roomRequestDto)
+          : await updateRoom(String(formData.id), roomRequestDto);
```
Comment (danieldietzler) [inline web/src/routes/admin/(resources)/rooms/[id]/+page.svelte:21]: Same here.

File: web/src/routes/admin/(resources)/rooms/[id]/+page.ts [hunk 6]
```diff
@@ -1,32 +1,45 @@
...
+    title = 'Raum — Hinzufügen';
+  } else {
+    try {
+      const room = await getRoom(params.id);
+      title = `Raum — ${room.name}`;
+      formRoom = {
+        id: room.id,
+        name: room.name,
+        buildingName: room.buildingName,
+        capacity: room.capacity,
```
Comment (danieldietzler) [inline web/src/routes/admin/(resources)/rooms/[id]/+page.ts:31]: Replace this part with `...room`

File: web/src/routes/admin/(resources)/studentGroups/[id]/+page.svelte [hunk 7]
```diff
@@ -1,43 +1,25 @@
...
-      studentIds: studentGroup.students.map((student) => student.id),
-      tagIds,
-      subjectIds: subjectIds || [],
-    };
-    try {
-      await updateStudentGroup(studentGroup.id, studentGroupRequestDto);
+      const _ =
+        formData.id === 'new'
+          ? await createStudentGroup(studentGroupRequestDto)
+          : await updateStudentGroup(String(formData.id), studentGroupRequestDto);
```
Comment (danieldietzler) [inline web/src/routes/admin/(resources)/studentGroups/[id]/+page.svelte:22]: Again, don't do `const _ =`

File: web/src/routes/admin/(resources)/students/[id]/+page.svelte [hunk 8]
```diff
@@ -1,48 +1,28 @@
...
-  async function update(values: string[], tagIds: string[]) {
-    let studentRequestDto: StudentRequestDto = {
-      firstName: values[0],
-      lastName: values[1],
-      groupIds: [],
-      tagIds,
-    };
-    try {
-      await updateStudent(student.id, studentRequestDto);
+      const _ = //eslint complains otherwise (no-unused-expressions)
```
Comment (danieldietzler) [inline web/src/routes/admin/(resources)/students/[id]/+page.svelte:18]: You even comment on it, but don't fix it properly lmao

File: web/src/routes/admin/(resources)/students/[id]/+page.ts [hunk 9]
```diff
@@ -1,32 +1,43 @@
...
-    error(404, { message: `Student with id ${params.id} not found` });
+    formStudent = { id: 'new', firstName: '', lastName: '', tags: [] };
+    title = 'Schüler — Hinzufügen';
+  } else {
+    try {
+      const student = await getStudent(params.id);
+      formStudent = {
+        id: student.id,
+        firstName: student.firstName,
+        lastName: student.lastName,
```
Comment (danieldietzler) [inline web/src/routes/admin/(resources)/students/[id]/+page.ts:28]: `...student`

File: web/src/routes/admin/(resources)/subjects/[id]/+page.svelte [hunk 10]
```diff
@@ -1,44 +1,26 @@
...
-  }
-
-  async function update(values: string[], tagIds: string[]) {
-    let subjectRequestDto: SubjectRequestDto = {
-      name: values[0],
-      tagIds,
-    };
-    try {
-      await updateSubject(subject.id, subjectRequestDto);
+      const _ =
```
Comment (danieldietzler) [inline web/src/routes/admin/(resources)/subjects/[id]/+page.svelte:16]: And here.

File: web/src/routes/admin/(resources)/tags/[id]/+page.svelte [hunk 11]
```diff
@@ -1,32 +1,21 @@
...
-      error(400, { message: 'Could not create tag' });
-    }
-  }
-
-  async function update(values: string[], _: string[]) {
-    //two args needed in component
-    let tagRequestDto: TagRequestDto = { tagName: values[0] };
-    try {
-      await updateTag(tag.id, tagRequestDto);
+      const _ =
```
Comment (danieldietzler) [inline web/src/routes/admin/(resources)/tags/[id]/+page.svelte:13]: .

File: web/src/routes/admin/(resources)/teachers/[id]/+page.svelte [hunk 12]
```diff
@@ -1,54 +1,39 @@
...
+      firstName: String(formData.firstName),
+      lastName: String(formData.lastName),
+      acronym: String(formData.acronym),
       subjectIds: subjectIds || [],
-      tagIds,
+      tagIds: formData.tags as string[],
     };
     try {
-      await createTeacher(teacherRequestDto);
+      const _ =
```
Comment (danieldietzler) [inline web/src/routes/admin/(resources)/teachers/[id]/+page.svelte:19]: .

File: web/src/routes/admin/(resources)/teachers/[id]/+page.ts [hunk 13]
```diff
@@ -1,35 +1,50 @@
...
+  } else {
+    try {
+      const teacher = await getTeacher(params.id);
+      title = `Teacher — ${teacher.firstName} ${teacher.lastName} (${teacher.acronym})`;
+      teacherSubjects = teacher.subjects.map((subject) => subject.id);
+      formTeacher = {
+        id: teacher.id,
+        firstName: teacher.firstName,
+        lastName: teacher.lastName,
+        acronym: teacher.acronym,
```
Comment (danieldietzler) [inline web/src/routes/admin/(resources)/teachers/[id]/+page.ts:34]: `...teacher`

File: web/src/routes/admin/(resources)/grades/[id]/+page.svelte (hunk 3 above)
Comment (Lama64) [reply web/src/routes/admin/(resources)/grades/[id]/+page.svelte:14]: I don't think typing `formData` stronger would be worth the effort.

Comment (danieldietzler) [reply web/src/routes/admin/(resources)/grades/[id]/+page.svelte:14]: Most likely at this point tbh

File: web/src/lib/components/ui/tags-multiple-select/tags-multiple-select.svelte [hunk 14]
```diff
@@ -2,17 +2,14 @@
...
+  export let selectedTagIds: string[];
 
-  onMount(() => {
-    selectedTags = entityTags.map((tag) => ({ label: tag.name, value: tag.id }));
-    selectedTagIds = selectedTags.map((tag) => tag.value);
-  });
+  let selectedTags: Selected<string>[] = tags
+    .filter((tag) => selectedTagIds.includes(tag.id))
+    .map((tag) => ({ label: tag.name, value: tag.id }));
+  selectedTagIds = selectedTags.map((tag) => tag.value);
```
Comment (danieldietzler) [inline web/src/lib/components/ui/tags-multiple-select/tags-multiple-select.svelte:12]: You can make this reactive, so replace the `let selectedTagIds` declaration with `$: selectedTagIds = selectedTags.map((tag) => tag.value)` and then you can also get rid of the `onSelectedChange` on the select below.

File: web/src/routes/admin/(resources)/grades/[id]/+page.ts [hunk 15]
```diff
@@ -1,39 +1,47 @@
...
+  id: z.string(),
+  name: z.string().trim().min(1, { message: 'Der Name darf nicht leer sein.' }),
+  tags: z.string().array(),
+  studentGroups: z.string().array(),
+});
+
 export const load = (async ({ params }) => {
   const sort: Sort = { sort: ['name,asc'] };
   const tags = await getTags(sort);
+  let formGrade, title;
```
Comment (danieldietzler) [inline web/src/routes/admin/(resources)/grades/[id]/+page.ts:18]: Those should have types

File: web/src/routes/admin/(resources)/rooms/[id]/+page.ts [hunk 16]
```diff
@@ -1,32 +1,42 @@
...
+  name: z.string().trim().min(1, { message: 'Der Name darf nicht leer sein.' }),
+  buildingName: z.string().trim().min(1, { message: 'Der Gebäudename darf nicht leer sein.' }),
+  capacity: z.coerce.number().int().positive({ message: 'Die Kapazität darf nicht 0 sein' }),
+  tags: z.string().array(),
+});
+
 export const load = (async ({ params }) => {
   const sort: Sort = { sort: ['name,asc'] };
   const tags = await getTags(sort);
+  let formRoom, title;
```
Comment (danieldietzler) [inline web/src/routes/admin/(resources)/rooms/[id]/+page.ts:19]: Here as well

File: web/src/routes/admin/(resources)/studentGroups/[id]/+page.ts [hunk 17]
```diff
@@ -1,52 +1,51 @@
...
+  students: z.string().array(),
+  tags: z.string().array(),
+});
+
 export const load = (async ({ params }) => {
   const sort: Sort = { sort: ['name,asc'] };
   const tags = await getTags(sort);
   const subjects = await getSubjects(sort);
   const grades = await getGrades(sort);
+  let formGroup, title, groupSubjects: string[];
```
Comment (danieldietzler) [inline web/src/routes/admin/(resources)/studentGroups/[id]/+page.ts:21]: And here

File: web/src/routes/admin/(resources)/students/[id]/+page.ts [hunk 18]
```diff
@@ -1,32 +1,41 @@
...
+  id: z.string(),
+  firstName: z.string().trim().min(1, { message: 'Der Vorname darf nicht leer sein.' }),
+  lastName: z.string().trim().min(1, { message: 'Der Nachname darf nicht leer sein.' }),
+  tags: z.string().array(),
+});
+
 export const load = (async ({ params }) => {
   const sort: Sort = { sort: ['name,asc'] };
   const tags = await getTags(sort);
+  let formStudent, title;
```
Comment (danieldietzler) [inline web/src/routes/admin/(resources)/students/[id]/+page.ts:18]: Type

File: web/src/routes/admin/(resources)/subjects/[id]/+page.ts [hunk 19]
```diff
@@ -1,32 +1,37 @@
...
+export const _schema = z.object({
+  id: z.string(),
+  name: z.string().trim().min(1, { message: 'Der Name darf nicht leer sein.' }),
+  tags: z.string().array(),
+});
+
 export const load = (async ({ params }) => {
   const sort: Sort = { sort: ['name,asc'] };
   const tags = await getTags(sort);
+  let formSubject, title;
```
Comment (danieldietzler) [inline web/src/routes/admin/(resources)/subjects/[id]/+page.ts:17]: Type

File: web/src/routes/admin/(resources)/teachers/[id]/+page.ts [hunk 20]
```diff
@@ -1,35 +1,47 @@
...
+  lastName: z.string().trim().min(1, { message: 'Der Nachname darf nicht leer sein.' }),
+  acronym: z.string().trim().min(1, { message: 'Das Akroynm darf nicht leer sein.' }),
+  tags: z.string().array(),
+});
+
 export const load = (async ({ params }) => {
   const sort: Sort = { sort: ['name,asc'] };
   const tags = await getTags(sort);
   const subjects = await getSubjects(sort);
+  let formTeacher, title, teacherSubjects: string[];
```
Comment (danieldietzler) [inline web/src/routes/admin/(resources)/teachers/[id]/+page.ts:20]: Types for `formTeacher` and `title`

**Review by danieldietzler — COMMENTED** [review]
A couple of final things before it's good to go :)

**Review by danieldietzler — APPROVED** [review]
LGTM

**Pull request merged**